SSA/Projects/RISCV64/Semantics2.lean

```diff
@@ -0,0 +1,642 @@
+/- we import these specific mathlib libraries to use Nat.cast_ofNat and neg_neg lemmas. -/
+import Mathlib.Order.Lattice
+import Mathlib.Data.Nat.Cast.Defs
+
+/-!
+  ## Dialect semantics bit-vectorization
+This file contains the `RV64` instruction semantics in a form that does
+not contain BitVec.toInt and BitVec.toNat. This file should replace
+the Semantics file in RISCV64 when complete.
+ToInt/ToNat calls hinder proof automation by bv_decide.
+To guarantee faithfullness to the semantics defined by the Sail-Lean RiscV64 model,
+we provide a proof with each rewrite that shows, that the toInt/ toNat free
+version is semantically equivalent to the defintion extracted from
+the Sail RiscV 64 modell in Lean.
+
+
+The instruction semantics suffixed by "bv" indicates that it is
+the purely bit vector defintion where we removed
+toInt/toNat calls.
+-/
+
+/-! ## Helper Bit Vector Lemmas  -/
+/- The lemmas below are helper lemmas used within proofs throughout the file. -/
+
+theorem sshiftRight_eq_sshiftRight_extractLsb {w : Nat}
+    {lw : Nat} (hll : 2^lw = w) (hlw : lw > 0 )
+    (x y : BitVec w)(h : BitVec.toNat y < 2^lw ) :
+        x.sshiftRight y.toNat = x.sshiftRight (y.extractLsb (lw - 1) 0).toNat := by
+  apply BitVec.eq_of_toNat_eq
+  simp
+  suffices y.toNat = (y.toNat % 2 ^ (lw - 1 + 1)) by
+    rw [← this ]
+  have : (lw - 1 + 1) = lw := by omega
+  rw [this, hll]
+  rw [Nat.mod_eq_of_lt]
+  rw [hll] at h
+  exact h
+
+theorem sshiftRight_eq_setWidth_extractLsb_signExtend {w : Nat} (n : Nat) (x : BitVec w) :
+    x.sshiftRight n =
+    ((x.signExtend (w + n)).extractLsb (w - 1 + n) n).setWidth w := by
+  ext i hi
+  simp [BitVec.getElem_sshiftRight]
+  simp [show i ≤ w - 1 by omega]
+  simp [BitVec.getLsbD_signExtend]
```

Comment:
```suggestion
  simp [BitVec.getElem_sshiftRight, show i ≤ w - 1 by omega, BitVec.getLsbD_signExtend]
```
if possible we should try to group things wihtin the same `simp` or `rw` (not always possible, so please do double check that this does not break the proof!) also, non-terminal simp